src/nmr_to_structure/prepare_input/prepare_nmr_rxn_input.py

```diff
@@ -1,92 +1,163 @@
+    token_space: str = "shared",
+    non_matching=True,
+    tanimoto_bins: int = 5,
+) -> pd.DataFrame:
+    all_smiles = list(nmr_df.index)
+    random.shuffle(all_smiles)
+
+    smiles_chunks = [
+        all_smiles[
+            i
+            * len(all_smiles)
+            // cores : min((i + 1) * len(all_smiles) // cores, len(all_smiles))
+        ]
+        for i in range(cores)
+    ]
+
+    manager = Manager()
+    return_list = manager.list()
+    status_list = manager.list([0] * len(smiles_chunks))
+    jobs = []
+
+    worker_fn = partial(
+        tanimoto_set_worker,
+        nmr_df=nmr_df,
+        mode=mode,
+        n_samples=n_samples // cores,
+        return_list=return_list,
+        status_list=status_list,
+        hnmr_mode=hnmr_mode,
+        token_space=token_space,
+        non_matching=non_matching,
+        tanimoto_bins=tanimoto_bins,
+    )
 
-        cnmr_string = build_cnmr_string(
-            cnmr,
-            header=False,
-            token_space=token_space,
-        )[0]
+    for i, chunk in enumerate(smiles_chunks):
+        logger.info(f"Starting job {i+1}/{cores}")
+        process = Process(target=worker_fn, args=(i, chunk))
+        jobs.append(process)
+        process.start()
 
-        nmr_string = " {} {}".format(hnmr_string.strip(), cnmr_string)
+    pbar = tqdm(total=n_samples)
+    while len(return_list) != len(smiles_chunks):
+        status = 0
+        for value in status_list:
+            status += value
+        pbar.update(status - pbar.n)
+        time.sleep(5)
+    pbar.close()
 
-    elif mode == "hnmr":
-        hnmr_string = build_hnmr_string(
-            smiles=component,
-            peak_dict=hnmr,
-            mode=hnmr_mode,
-            header=False,
-            token_space=token_space,
-        )[0]
+    combined_df = pd.concat(return_list)
 
-        nmr_string = " " + hnmr_string
+    logger.info("Printing Tanimoto statistics:")
+    for key in pd.unique(combined_df.tanimoto_bin):
+        logger.info(
+            "Tanimoto bin {}: {:.3f}%".format(
+                key, (combined_df.tanimoto_bin == key).sum() / len(combined_df) * 100
+            )
+        )
 
-    elif mode == "cnmr":
-        cnmr_string = build_cnmr_string(
-            cnmr,
-            header=False,
-            token_space=token_space,
-        )[0]
-        nmr_string = " " + cnmr_string
-    
-    return nmr_string
+    return combined_df
 
 
 def make_nmr_rxn_set(
-    nmr_set: pd.DataFrame,
+    nmr_df: pd.DataFrame,
     rxn_set: pd.DataFrame,
+    n_samples: int,
     mode: str,
     hnmr_mode: str = "range",
-    token_space: str ="shared",
-    non_matching: bool = True
-
+    token_space: str = "shared",
+    non_matching: bool = True,
 ) -> pd.DataFrame:
     rxns = list(rxn_set.index)
-    mols = set(nmr_set.index)
+    random.shuffle(rxns)
+
+    mols = set(nmr_df.index)
     mols_sample = tuple(mols)
 
-    src_tgt_pairs = list()
+    src_tgt_pairs: List[dict] = list()
```

Comment:
Do you need the content to be a dictionary? If not, `List[Tuple[str, str]]` may be more appropriate.